base/abstractarray.jl

```diff
@@ -98,6 +98,16 @@ axes1(iter) = OneTo(length(iter))
 unsafe_indices(A) = axes(A)
 unsafe_indices(r::AbstractRange) = (OneTo(unsafe_length(r)),) # Ranges use checked_sub for size
 
+"""
+    keys(a::AbstractArray)
+
+Return the tuple of valid indices for array `a`. This is equivalent to
+[`LinearIndices(a)`](@ref) for vectors, and to [`CartesianIndices(axes(a))`](@ref)
+for higher-dimensional arrays.
+
+This may not return the most efficient indices type to iterate over `a`:
+use [`eachindex`](@ref) instead for maximum performance.
```

Comment:
I don't think we need to explicitly demand that custom arrays return either `LinearIndices` or `CartesianIndices` — the important part is simply that they're arrays of integers or `CartesianIndex`.  I do like the pointers to them, though.  I think I'd prefer to see this written more like:

```suggestion
Return an efficient array describing all valid indices for `a` arranged in the shape of `a` itself.

They keys of 1-dimensional arrays (vectors) are integers, whereas all other N-dimensional
arrays use [`CartesianIndex`es](@ref) to describe their locations.  Often the special array
types [`LinearIndices`](@ref) and [`CartesianIndices`](@ref) are used to efficiently
represent these arrays of integers and `CartesianIndex`es, respectively.

Note that the `keys` of an array might not be the most efficient index type; for maximum
performance use  [`eachindex`](@ref) instead.
```